R/standardize_treatment.R

```diff
@@ -0,0 +1,445 @@
+                       as.numeric(nj), lambda, lowlim, uplim, data_in)
+
+  # create propensity score multipliers and split by site
+  pro_trt <- Z / (pscores * nj[S_factor])
+  pro_trt_split <- split(pro_trt, S_factor)
+  pro_ctr <- (1 - Z) / ((1 - pscores) * nj[S_factor])
+  pro_ctr_split <- split(pro_ctr, S_factor)
+
+  # define number of auxiliary weights
+  aux_dim <- (J * ncol(X0)) + (J * ncol(Xtau))
+
+  # construct linear term vector
+  if(verbose) message("Creating linear term vector...")
+  if(is.null(data_in$q)) {
+    q <- create_q_vector_treatment(Xtaus, pro_trt_split, target, aux_dim)
+  } else {
+    q <- data_in$q
+  }
+
+  # construct quadratic term matrix
+  if(verbose) message("Creating quadratic term matrix...")
+  if(is.null(data_in$P)) {
+    P <- create_P_matrix_treatment(n, aux_dim)
+  } else {
+    P <- data_in$P
+  }
+  I0 <- create_I0_matrix_treatment(pro_trt_split, pro_ctr_split,
+                                   scale_sample_size, nj, n, aux_dim)
+  P <- P + lambda * I0
+
+  # construct constraint matrix
+  if(verbose) message("Creating constraint matrix...")
+  if(is.null(data_in$constraints)) {
+    constraints <- create_constraints_treatment(X0s, Xtaus, target, Z, S_factor,
+                                                pro_trt_split, pro_ctr_split,
+                                                lowlim, uplim, exact_global, verbose)
+  } else {
+    constraints <- data_in$constraints
+    constraints$l[(J + 1):(J + n)] <- lowlim
+    constraints$u[(J + 1):(J + n)] <- uplim
+  }
+
+  # set optimization settings
+  if(verbose) {
+    settings <- osqp::osqpSettings(verbose = TRUE, eps_abs = 1e-6,
+                                   eps_rel = 1e-6, max_iter = 5000)
+  } else {
+    settings <- osqp::osqpSettings(verbose = FALSE, eps_abs = 1e-6,
+                                   eps_rel = 1e-6, max_iter = 5000)
+  }
+  settings <- do.call(osqp::osqpSettings,
+                      c(list(verbose = verbose,
+                             eps_rel = eps_rel,
+                             eps_abs = eps_abs),
+                        list(...)))
+
+  # solve optimization problem (possibly with uniform weights)
+  if(init_uniform) {
+    if(verbose) message("Initializing with uniform weights")
+    unifw <- get_uniform_weights_treatment(X0s, Xtaus)
+    obj <- osqp::osqp(P, q, constraints$A,
+                      constraints$l, constraints$u, pars = settings)
+    obj$WarmStart(x = unifw)
+    solution <- obj$Solve()
+  } else {
+    solution <- osqp::solve_osqp(P, q, constraints$A,
+                                 constraints$l, constraints$u,
+                                 pars = settings)
+  }
+
+  # convert weights into a matrix
+  if(verbose) message("Reordering weights...")
+  weights <- matrix(0, ncol = J, nrow = n)
+  cumsumnj <- cumsum(c(1, nj))
+  for(j in 1:J) {
+    weights[idxs[[j]], j] <- solution$x[cumsumnj[j]:(cumsumnj[j + 1] - 1)]
+  }
```

Comment:
I think it probably makes sense to keep the weights as a vector. I know that the standardize code doesn't do that, but checkout [here ](https://github.com/ebenmichael/balancer/blob/f20777fd60d98c7dfb4f5c49b7740350dad2116a/R/multilevel_qp.R#L77-L85) for code that does.

Reply:
Done.